realm/realm-library/src/main/cpp/java_binding_context.hpp

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2016 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JAVA_BINDING_CONTEXT_HPP
+#define JAVA_BINDING_CONTEXT_HPP
+
+#include <jni.h>
+#include <memory>
+
+#include "binding_context.hpp"
+
+namespace realm {
+
+namespace _impl {
+
+// Binding context which will be called from OS.
+class JavaBindingContext final : public BindingContext {
+private:
+    struct ConcreteJavaBindContext {
```

Comment:
To make the `JavaBindingContext` constructor private and still be able to use `std::make_shared`. See http://stackoverflow.com/questions/8147027/how-do-i-call-stdmake-shared-on-a-class-with-only-protected-or-private-const/8147213#8147213